src/beast/beast/http/headers.h

```diff
@@ -49,170 +52,218 @@ class headers
+
+    /** Move assignment.
+
+        After the move, the moved-from object has an empty list of fields.
+    */
+    headers& operator=(headers&&);
 
-    headers (headers&& other);
-    headers& operator= (headers&& other);
+    /** Copy constructor. */
+    headers(headers const&);
 
-    headers (headers const& other);
-    headers& operator= (headers const& other);
+    /** Copy assignment. */
+    headers& operator=(headers const&);
 
-    /** Returns an iterator to headers in order of appearance. */
+    /** Returns an iterator to fields in order of appearance. */
     /** @{ */
     iterator
-    begin() const;
+    begin() const
+    {
+        return {list_.cbegin(), transform{}};
+    }
 
     iterator
-    end() const;
+    end() const
+    {
+        return {list_.cend(), transform{}};
+    }
 
     iterator
-    cbegin() const;
+    cbegin() const
+    {
+        return {list_.cbegin(), transform{}};
+    }
 
     iterator
-    cend() const;
+    cend() const
+    {
+        return {list_.cend(), transform{}};
+    }
     /** @} */
 
+    /** Returns `true` if the specified field exists. */
+    bool
+    exists(std::string const& field) const
+    {
+        return set_.find(field, less{}) != set_.end();
+    }
+
     /** Returns an iterator to the case-insensitive matching header. */
-    template <class = void>
     iterator
-    find (std::string const& field) const;
+    find(std::string const& field) const;
 
     /** Returns the value for a case-insensitive matching header, or "" */
-    template <class = void>
     std::string const&
-    operator[] (std::string const& field) const;
+    operator[](std::string const& field) const;
 
     /** Clear the contents of the headers. */
-    template <class = void>
     void
     clear() noexcept;
 
     /** Remove a field.
+
         @return The number of fields removed.
     */
-    template <class = void>
     std::size_t
-    erase (std::string const& field);
+    erase(std::string const& field);
+
+    /** Insert a field value.
 
-    /** Append a field value.
         If a field value already exists the new value will be
         extended as per RFC2616 Section 4.2.
     */
-    // VFALCO TODO Consider allowing rvalue references for std::move
-    template <class = void>
+    // VFALCO TODO Consider allowing rvalue references for std::move?
     void
-    append (std::string const& field, std::string const& value);
-};
+    insert(std::string const& field, std::string const& value);
 
-template <class = void>
-std::string
-to_string (headers const& h);
+    /** Insert a field value.
 
-// HACK!
-template <class = void>
-std::map <std::string, std::string>
-build_map (headers const& h);
+        If a field value already exists the new value will be
+        extended as per RFC2616 Section 4.2.
+    */
+    template<class T,
+        class = std::enable_if_t<
+            ! std::is_constructible<std::string, T>::value>>
+    void
+    insert(std::string const& field, T&& t)
```

Comment:
Since `boost::lexical_cast` takes it's parameter by `const&` forwarding ref doesn't give any benefit here. I'd change to `const&` and get rid of the `std::forward`.